File: CUDA/wall_fluxes_cuda.jl
function wall_dGx_pos_kernel(gpuGlobalDataConn, gpuGlobalDataFixedPoint, gpuGlobalDataFauxFixed, gpuGlobalDataRest, gpuConfigData, numPoints, shared, flux_shared, qtilde_shared)
    idx = (blockIdx().x - 1) * blockDim().x + threadIdx().x
    thread_idx = threadIdx().x
    block_dim = blockDim().x

    sum_delx_sqr = 0.0
    sum_dely_sqr = 0.0
    sum_delx_dely = 0.0
    sum_delx_delf = SVector{4,Float64}(0, 0, 0, 0)
    sum_dely_delf = SVector{4,Float64}(0, 0, 0, 0)


    x_i = gpuGlobalDataFauxFixed[idx]
    y_i = gpuGlobalDataFauxFixed[idx + numPoints]
    nx = gpuGlobalDataFauxFixed[idx + 2 * numPoints]
    ny = gpuGlobalDataFauxFixed[idx + 3 * numPoints]


    power = gpuConfigData[6]
    # gamma = gpuConfigData[15]
    for iter in 15:24
        conn = gpuGlobalDataConn[idx, iter]
        if conn == 0.0
            break
        end

        # x_k = gpuGlobalDataFixedPoint[conn].x
        # y_k = gpuGlobalDataFixedPoint[conn].y
        delx = gpuGlobalDataFauxFixed[conn] - x_i
        dely = gpuGlobalDataFauxFixed[conn + numPoints] - y_i
        dels = delx*ny - dely*nx
        deln = delx*nx + dely*ny
        dist = CUDAnative.hypot(dels, deln)
        weights = CUDAnative.pow(dist, power)
        # weights = 1.0


        dels_weights = dels*weights
        deln_weights = deln*weights

        sum_delx_sqr += dels*dels_weights
        sum_dely_sqr += deln*deln_weights
        sum_delx_dely += dels*deln_weights

        # if idx == 3
        #     @cuprintf("\n %d", conn)
        #     @cuprintf("\n %.17f %.17f %.17f %.17f", qtilde_k[1], qtilde_k[2], qtilde_k[3], qtilde_k[4])
        # end

        shared[thread_idx], shared[thread_idx + block_dim], shared[thread_idx + block_dim * 2], shared[thread_idx + block_dim * 3] = 0.0, 0.0, 0.0, 0.0
        venkat_limiter_kernel_qtilde(gpuGlobalDataFixedPoint, gpuGlobalDataRest, idx, gpuConfigData, delx, dely, shared, qtilde_shared)
        flux_quad_GxII_kernel(nx, ny, idx, shared, +)
        venkat_limiter_kernel_qtilde(gpuGlobalDataFixedPoint, gpuGlobalDataRest, conn, gpuConfigData, delx, dely, shared, qtilde_shared)
        flux_quad_GxII_kernel(nx, ny, idx, shared, -)

        temp_var = @SVector [shared[thread_idx], shared[thread_idx + block_dim], shared[thread_idx + block_dim * 2], shared[thread_idx + block_dim * 3]]
        sum_delx_delf += temp_var * dels_weights
        sum_dely_delf += temp_var * deln_weights

    end

    det = sum_delx_sqr*sum_dely_sqr - sum_delx_dely*sum_delx_dely
    one_by_det = 1.0 / det
    # if idx ==3
    #     @cuprintf("\n===Gn===")
    #     @cuprintf("\n %.17f %.17f %.17f %.17f", sum_1,sum_2,sum_3,sum_4)
    #     @cuprintf("\n %.17f %.17f %.17f %.17f", sum_5,sum_6,sum_7,sum_8)
    # end
    flux_shared[thread_idx] += 2 * (sum_delx_delf[1]*sum_dely_sqr - sum_dely_delf[1]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim] += 2 * (sum_delx_delf[2]*sum_dely_sqr - sum_dely_delf[2]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim * 2] += 2 * (sum_delx_delf[3]*sum_dely_sqr - sum_dely_delf[3]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim * 3] += 2 * (sum_delx_delf[4]*sum_dely_sqr - sum_dely_delf[4]*sum_delx_dely)*one_by_det
    # if idx ==3
    #     @cuprintf("\n %.17f %.17f %.17f %.17f", gpuGlobalDataRest[5, idx],gpuGlobalDataRest[6, idx],gpuGlobalDataRest[7, idx],gpuGlobalDataRest[8, idx])
    # end
    return nothing
end

function wall_dGx_neg_kernel(gpuGlobalDataConn, gpuGlobalDataFixedPoint, gpuGlobalDataFauxFixed, gpuGlobalDataRest, gpuConfigData, numPoints, shared, flux_shared, qtilde_shared)
    idx = (blockIdx().x - 1) * blockDim().x + threadIdx().x
    thread_idx = threadIdx().x
    block_dim = blockDim().x

    sum_delx_sqr = 0.0
    sum_dely_sqr = 0.0
    sum_delx_dely = 0.0
    sum_delx_delf = SVector{4,Float64}(0, 0, 0, 0)
    sum_dely_delf = SVector{4,Float64}(0, 0, 0, 0)
    # shared[thread_idx + block_dim * 8], shared[thread_idx + block_dim * 9], shared[thread_idx + block_dim * 10], shared[thread_idx + block_dim * 11] = 0.0, 0.0, 0.0, 0.0
    # shared[thread_idx + block_dim * 12], shared[thread_idx + block_dim * 13], shared[thread_idx + block_dim * 14], shared[thread_idx + block_dim * 15] = 0.0, 0.0, 0.0, 0.0
    x_i = gpuGlobalDataFauxFixed[idx]
    y_i = gpuGlobalDataFauxFixed[idx + numPoints]
    nx = gpuGlobalDataFauxFixed[idx + 2 * numPoints]
    ny = gpuGlobalDataFauxFixed[idx + 3 * numPoints]


    power = gpuConfigData[6]
    # gamma = gpuConfigData[15]
    for iter in 25:34
        conn = gpuGlobalDataConn[idx, iter]
        if conn == 0.0
            break
        end

        # x_k = gpuGlobalDataFixedPoint[conn].x
        # y_k = gpuGlobalDataFixedPoint[conn].y
        delx = gpuGlobalDataFauxFixed[conn] - x_i
        dely = gpuGlobalDataFauxFixed[conn + numPoints] - y_i
        dels = delx*ny - dely*nx
        deln = delx*nx + dely*ny
        dist = CUDAnative.hypot(dels, deln)
        weights = CUDAnative.pow(dist, power)
        # weights = 1.0


        dels_weights = dels*weights
        deln_weights = deln*weights

        sum_delx_sqr += dels*dels_weights
        sum_dely_sqr += deln*deln_weights
        sum_delx_dely += dels*deln_weights

        shared[thread_idx], shared[thread_idx + block_dim], shared[thread_idx + block_dim * 2], shared[thread_idx + block_dim * 3] = 0.0, 0.0, 0.0, 0.0
        venkat_limiter_kernel_qtilde(gpuGlobalDataFixedPoint, gpuGlobalDataRest, idx, gpuConfigData, delx, dely, shared, qtilde_shared)
        flux_quad_GxI_kernel(nx, ny, idx, shared, +)
        venkat_limiter_kernel_qtilde(gpuGlobalDataFixedPoint, gpuGlobalDataRest, conn, gpuConfigData, delx, dely, shared, qtilde_shared)
        flux_quad_GxI_kernel(nx, ny, idx, shared, -)
        # CUDAnative.synchronize()
        temp_var = @SVector [shared[thread_idx], shared[thread_idx + block_dim], shared[thread_idx + block_dim * 2], shared[thread_idx + block_dim * 3]]
        sum_delx_delf += temp_var * dels_weights
        sum_dely_delf += temp_var * deln_weights

        # shared[thread_idx + block_dim * 8], shared[thread_idx + block_dim * 9], shared[thread_idx + block_dim * 10], shared[thread_idx + block_dim * 11] = shared[thread_idx]* dels_weights,
        # shared[thread_idx + block_dim]* dels_weights, shared[thread_idx + block_dim * 2]* dels_weights, shared[thread_idx + block_dim * 3]* dels_weights

        # shared[thread_idx + block_dim * 12], shared[thread_idx + block_dim * 13], shared[thread_idx + block_dim * 14], shared[thread_idx + block_dim * 15] = shared[thread_idx]* deln_weights,
        # shared[thread_idx + block_dim]* deln_weights, shared[thread_idx + block_dim * 2]* deln_weights, shared[thread_idx + block_dim * 3]* deln_weights

        # if idx == 3
        #     @cuprintf("\n *** ")
        #     @cuprintf("\n %f %f %f %f", gpuGlobalDataRest[37, idx], gpuGlobalDataRest[38, idx], gpuGlobalDataRest[39, idx], gpuGlobalDataRest[40, idx])
        #     @cuprintf("\n %f %f %f %f", gpuGlobalDataRest[41, idx], gpuGlobalDataRest[42, idx], gpuGlobalDataRest[43, idx], gpuGlobalDataRest[44, idx])
        #     @cuprintf("\n %f", dels_weights)
        #     @cuprintf("\n %f", deln_weights)
        # end
    end

    det = sum_delx_sqr*sum_dely_sqr - sum_delx_dely*sum_delx_dely
    one_by_det = 1.0 / det

    flux_shared[thread_idx] += 2 * (sum_delx_delf[1]*sum_dely_sqr - sum_dely_delf[1]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim] += 2 * (sum_delx_delf[2]*sum_dely_sqr - sum_dely_delf[2]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim * 2] += 2 * (sum_delx_delf[3]*sum_dely_sqr - sum_dely_delf[3]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim * 3] += 2 * (sum_delx_delf[4]*sum_dely_sqr - sum_dely_delf[4]*sum_delx_dely)*one_by_det

    # flux_shared[thread_idx] += 2 * (shared[thread_idx + block_dim * 8]*sum_dely_sqr - shared[thread_idx + block_dim * 12]*sum_delx_dely)*one_by_det
    # flux_shared[thread_idx + block_dim] += 2 * (shared[thread_idx + block_dim * 9]*sum_dely_sqr - shared[thread_idx + block_dim * 13]*sum_delx_dely)*one_by_det
    # flux_shared[thread_idx + block_dim * 2] += 2 * (shared[thread_idx + block_dim * 10]*sum_dely_sqr - shared[thread_idx + block_dim * 14]*sum_delx_dely)*one_by_det
    # flux_shared[thread_idx + block_dim * 3] += 2 * (shared[thread_idx + block_dim * 11]*sum_dely_sqr - shared[thread_idx + block_dim * 15]*sum_delx_dely)*one_by_det
    # if idx ==3
    #     @cuprintf("\n %f %f %f %f", gpuGlobalDataRest[5, idx],gpuGlobalDataRest[6, idx],gpuGlobalDataRest[7, idx],gpuGlobalDataRest[8, idx])
    # end
    return nothing
end

function wall_dGy_neg_kernel(gpuGlobalDataConn, gpuGlobalDataFixedPoint, gpuGlobalDataFauxFixed, gpuGlobalDataRest, gpuConfigData, numPoints, shared, flux_shared, qtilde_shared)
    idx = (blockIdx().x - 1) * blockDim().x + threadIdx().x
    thread_idx = threadIdx().x
    block_dim = blockDim().x

    sum_delx_sqr = 0.0
    sum_dely_sqr = 0.0
    sum_delx_dely = 0.0
    sum_delx_delf = SVector{4,Float64}(0, 0, 0, 0)
    sum_dely_delf = SVector{4,Float64}(0, 0, 0, 0)

    x_i = gpuGlobalDataFauxFixed[idx]
    y_i = gpuGlobalDataFauxFixed[idx + numPoints]
    nx = gpuGlobalDataFauxFixed[idx + 2 * numPoints]
    ny = gpuGlobalDataFauxFixed[idx + 3 * numPoints]

    power = gpuConfigData[6]
    # gamma = gpuConfigData[15]

    for iter in 45:54
        conn = gpuGlobalDataConn[idx, iter]
        if conn == 0.0
            break
        end

        # x_k = gpuGlobalDataFixedPoint[conn].x
        # y_k = gpuGlobalDataFixedPoint[conn].y
        delx = gpuGlobalDataFauxFixed[conn] - x_i
        dely = gpuGlobalDataFauxFixed[conn + numPoints] - y_i
        dels = delx*ny - dely*nx
        deln = delx*nx + dely*ny
        dist = CUDAnative.hypot(dels, deln)
        weights = CUDAnative.pow(dist, power)
        # weights = 1.0


        dels_weights = dels*weights
        deln_weights = deln*weights

        sum_delx_sqr += dels*dels_weights
        sum_dely_sqr += deln*deln_weights
        sum_delx_dely += dels*deln_weights

        shared[thread_idx], shared[thread_idx + block_dim], shared[thread_idx + block_dim * 2], shared[thread_idx + block_dim * 3] = 0.0, 0.0, 0.0, 0.0
        venkat_limiter_kernel_qtilde(gpuGlobalDataFixedPoint, gpuGlobalDataRest, idx, gpuConfigData, delx, dely, shared, qtilde_shared)
        flux_Gyn_kernel(nx, ny, idx, shared, +)
        venkat_limiter_kernel_qtilde(gpuGlobalDataFixedPoint, gpuGlobalDataRest, conn, gpuConfigData, delx, dely, shared, qtilde_shared)
        flux_Gyn_kernel(nx, ny, idx, shared, -)
        # CUDAnative.synchronize()
        temp_var = @SVector [shared[thread_idx], shared[thread_idx + block_dim], shared[thread_idx + block_dim * 2], shared[thread_idx + block_dim * 3]]
        sum_delx_delf += temp_var * dels_weights
        sum_dely_delf += temp_var * deln_weights

    end

    det = sum_delx_sqr*sum_dely_sqr - sum_delx_dely*sum_delx_dely
    one_by_det = 1.0 / det

    flux_shared[thread_idx] += 2 * (sum_dely_delf[1]*sum_delx_sqr - sum_delx_delf[1]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim] += 2 * (sum_dely_delf[2]*sum_delx_sqr - sum_delx_delf[2]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim * 2] += 2 * (sum_dely_delf[3]*sum_delx_sqr - sum_delx_delf[3]*sum_delx_dely)*one_by_det
    flux_shared[thread_idx + block_dim * 3] += 2 * (sum_dely_delf[4]*sum_delx_sqr - sum_delx_delf[4]*sum_delx_dely)*one_by_det

    return nothing
end

File: wall_fluxes_cuda.py
import math
import limiters_cuda
import split_fluxes_cuda
import quadrant_fluxes_cuda
import numba
from numba import cuda
from cuda_func import add, zeros, multiply, qtilde_to_primitive_cuda, subtract, multiply_element_wise_shared
from operator import add as addop, sub as subop

@cuda.jit(inline=True)
def wall_dGx_pos(x, y, nx_gpu, ny_gpu, flag_1_gpu, min_dist, nbhs, conn, xpos_nbhs, xpos_conn, prim, q, maxminq, dq, flux_res, power, vl_const, gamma, wall):

    tx = cuda.threadIdx.x
    bx = cuda.blockIdx.x
    bw = cuda.blockDim.x
    idx =  bx * bw + tx

    if flag_1_gpu[idx] != wall:
        return
    
    if idx == 0 or idx >= len(x):
        return

    other_shared = cuda.shared.array(shape = (1024 * 2), dtype=numba.float64)
    sum_delx_delf = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)
    sum_dely_delf = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)
    qtilde_shared = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)

    sum_delx_sqr = 0
    sum_dely_sqr = 0
    sum_delx_dely = 0

    x_i = x[idx]
    y_i = y[idx]

    nx = nx_gpu[idx]
    ny = ny_gpu[idx]

    tx = ny
    ty = -nx

    for i in range(4):
        sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] = 0
        sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] = 0

    for itm in xpos_conn[idx][:xpos_nbhs[idx]]:

        itm = conn[idx][itm]
        x_k = x[itm]
        y_k = y[itm]

        delx = x_k - x_i
        dely = y_k - y_i

        dels = delx*tx + dely*ty
        deln = delx*nx + dely*ny

        dist = (dels*dels + deln*deln) ** 0.5
        weights = dist**power

        dels_weights = dels*weights
        deln_weights = deln*weights

        sum_delx_sqr = sum_delx_sqr + dels*dels_weights
        sum_dely_sqr = sum_dely_sqr + deln*deln_weights

        sum_delx_dely = sum_delx_dely + dels*deln_weights

        other_shared[cuda.threadIdx.x], other_shared[cuda.threadIdx.x + cuda.blockDim.x], other_shared[cuda.threadIdx.x + cuda.blockDim.x * 2], other_shared[cuda.threadIdx.x + cuda.blockDim.x * 3] = 0, 0, 0, 0

        limiters_cuda.venkat_limiter(qtilde_shared, q, maxminq, dq, nbhs, conn, x, y, min_dist, idx, vl_const, other_shared, delx, dely, gamma)
        quadrant_fluxes_cuda.flux_quad_GxII(nx, ny, other_shared, addop)

        limiters_cuda.venkat_limiter(qtilde_shared, q, maxminq, dq, nbhs, conn, x, y, min_dist, itm, vl_const, other_shared, delx, dely, gamma)
        quadrant_fluxes_cuda.flux_quad_GxII(nx, ny, other_shared, subop)

        for i in range(4):
            sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] += dels_weights * other_shared[cuda.threadIdx.x + cuda.blockDim.x * i]
            sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] += deln_weights * other_shared[cuda.threadIdx.x + cuda.blockDim.x * i]

    det = sum_delx_sqr*sum_dely_sqr - sum_delx_dely*sum_delx_dely
    one_by_det = 1 / det

    for i in range(4):
        sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] *= sum_dely_sqr
        sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] *= sum_delx_dely

    for i in range(4):
        flux_res[idx][i] = 2 * one_by_det * (sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] - sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i])

@cuda.jit(inline=True)
def wall_dGx_neg(x, y, nx_gpu, ny_gpu, flag_1_gpu, min_dist, nbhs, conn, xneg_nbhs, xneg_conn, prim, q, maxminq, dq, flux_res, power, vl_const, gamma, wall):

    tx = cuda.threadIdx.x
    bx = cuda.blockIdx.x
    bw = cuda.blockDim.x
    idx =  bx * bw + tx

    if flag_1_gpu[idx] != wall:
        return
    
    if idx == 0 or idx >= len(x):
        return

    other_shared = cuda.shared.array(shape = (1024 * 2), dtype=numba.float64)
    sum_delx_delf = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)
    sum_dely_delf = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)
    qtilde_shared = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)

    sum_delx_sqr = 0
    sum_dely_sqr = 0
    sum_delx_dely = 0

    x_i = x[idx]
    y_i = y[idx]

    nx = nx_gpu[idx]
    ny = ny_gpu[idx]

    tx = ny
    ty = -nx

    for i in range(4):
        sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] = 0
        sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] = 0

    for itm in xneg_conn[idx][:xneg_nbhs[idx]]:

        itm = conn[idx][itm]
        x_k = x[itm]
        y_k = y[itm]

        delx = x_k - x_i
        dely = y_k - y_i

        dels = delx*tx + dely*ty
        deln = delx*nx + dely*ny

        dist = (dels*dels + deln*deln) ** 0.5
        weights = dist**power

        dels_weights = dels*weights
        deln_weights = deln*weights

        sum_delx_sqr = sum_delx_sqr + dels*dels_weights
        sum_dely_sqr = sum_dely_sqr + deln*deln_weights

        sum_delx_dely = sum_delx_dely + dels*deln_weights

        other_shared[cuda.threadIdx.x], other_shared[cuda.threadIdx.x + cuda.blockDim.x], other_shared[cuda.threadIdx.x + cuda.blockDim.x * 2], other_shared[cuda.threadIdx.x + cuda.blockDim.x * 3] = 0, 0, 0, 0

        limiters_cuda.venkat_limiter(qtilde_shared, q, maxminq, dq, nbhs, conn, x, y, min_dist, idx, vl_const, other_shared, delx, dely, gamma)
        quadrant_fluxes_cuda.flux_quad_GxI(nx, ny, other_shared, addop)

        limiters_cuda.venkat_limiter(qtilde_shared, q, maxminq, dq, nbhs, conn, x, y, min_dist, itm, vl_const, other_shared, delx, dely, gamma)
        quadrant_fluxes_cuda.flux_quad_GxI(nx, ny, other_shared, subop)

        for i in range(4):
            sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] += dels_weights * other_shared[cuda.threadIdx.x + cuda.blockDim.x * i]
            sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] += deln_weights * other_shared[cuda.threadIdx.x + cuda.blockDim.x * i]

    det = sum_delx_sqr*sum_dely_sqr - sum_delx_dely*sum_delx_dely
    one_by_det = 1 / det

    for i in range(4):
        sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] *= sum_dely_sqr
        sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] *= sum_delx_dely

    for i in range(4):
        flux_res[idx][i] += 2 * one_by_det * (sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] - sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i])

@cuda.jit(inline=True)
def wall_dGy_neg(x, y, nx_gpu, ny_gpu, flag_1_gpu, min_dist, nbhs, conn, yneg_nbhs, yneg_conn, prim, q, maxminq, dq, flux_res, power, vl_const, gamma, wall):

    tx = cuda.threadIdx.x
    bx = cuda.blockIdx.x
    bw = cuda.blockDim.x
    idx =  bx * bw + tx

    if flag_1_gpu[idx] != wall:
        return
    
    if idx == 0 or idx >= len(x):
        return

    other_shared = cuda.shared.array(shape = (1024 * 2), dtype=numba.float64)
    sum_delx_delf = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)
    sum_dely_delf = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)
    qtilde_shared = cuda.shared.array(shape = (256 * 2), dtype=numba.float64)

    sum_delx_sqr = 0
    sum_dely_sqr = 0
    sum_delx_dely = 0

    x_i = x[idx]
    y_i = y[idx]

    nx = nx_gpu[idx]
    ny = ny_gpu[idx]

    tx = ny
    ty = -nx

    for i in range(4):
        sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] = 0
        sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] = 0

    for itm in yneg_conn[idx][:yneg_nbhs[idx]]:

        itm = conn[idx][itm]
        x_k = x[itm]
        y_k = y[itm]

        delx = x_k - x_i
        dely = y_k - y_i

        dels = delx*tx + dely*ty
        deln = delx*nx + dely*ny

        dist = (dels*dels + deln*deln) ** 0.5
        weights = dist**power

        dels_weights = dels*weights
        deln_weights = deln*weights

        sum_delx_sqr = sum_delx_sqr + dels*dels_weights
        sum_dely_sqr = sum_dely_sqr + deln*deln_weights

        sum_delx_dely = sum_delx_dely + dels*deln_weights

        other_shared[cuda.threadIdx.x], other_shared[cuda.threadIdx.x + cuda.blockDim.x], other_shared[cuda.threadIdx.x + cuda.blockDim.x * 2], other_shared[cuda.threadIdx.x + cuda.blockDim.x * 3] = 0, 0, 0, 0

        limiters_cuda.venkat_limiter(qtilde_shared, q, maxminq, dq, nbhs, conn, x, y, min_dist, idx, vl_const, other_shared, delx, dely, gamma)
        split_fluxes_cuda.flux_Gyn(nx, ny, other_shared, addop)

        limiters_cuda.venkat_limiter(qtilde_shared, q, maxminq, dq, nbhs, conn, x, y, min_dist, itm, vl_const, other_shared, delx, dely, gamma)
        split_fluxes_cuda.flux_Gyn(nx, ny, other_shared, subop)

        for i in range(4):
            sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] += dels_weights * other_shared[cuda.threadIdx.x + cuda.blockDim.x * i]
            sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] += deln_weights * other_shared[cuda.threadIdx.x + cuda.blockDim.x * i]

    det = sum_delx_sqr*sum_dely_sqr - sum_delx_dely*sum_delx_dely
    one_by_det = 1 / det

    for i in range(4):
        sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i] *= sum_delx_dely
        sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] *= sum_delx_sqr

    for i in range(4):
        flux_res[idx][i] += 2 * one_by_det * (sum_dely_delf[cuda.threadIdx.x + cuda.blockDim.x * i] - sum_delx_delf[cuda.threadIdx.x + cuda.blockDim.x * i])